### PR TITLE
Add option to use an absolute healthcheck path

### LINKED ISSUE #531: add flag to allow healthcheck_path to be interpreted as either relative or absolute path
it can happen that you configure your proxy to redirect bulk requests into one path, but wish to trigger the health checks on a different path which is not relative off of the initial path.

we could add a boolean option that would interpret healthcheck_path as relative or absolute

### PR DESCRIPTION
**Pull request by suyograo**
Fixes #531

### CODE REVIEW

**Comment by suyograo** [comment]
@jsvd @andrewvc my rspec-foo is not great. can you check if I'm doing the right thing in these tests?

File: lib/logstash/outputs/elasticsearch/http_client/pool.rb [hunk 1]
```diff
@@ -225,9 +229,16 @@ def healthcheck!
...
       @state_mutex.synchronize { @url_info.select {|url,meta| meta[:state] != :alive } }.each do |url,meta|
         begin
-          logger.info("Running health check to see if an Elasticsearch connection is working",
-                      url: url, healthcheck_path: healthcheck_path)
-          response = perform_request_to_url(url, :head, healthcheck_path)
+          if @absolute_healthcheck_path
+            absolute_healthcheck_safe_url = ::LogStash::Util::SafeURI.new(healthcheck_path)
+            logger.info("Running health check to see if an Elasticsearch connection is working",
+                        absolute_healthcheck_url: absolute_healthcheck_safe_url.sanitized)
+            response = perform_request_to_url(healthcheck_path, :head, ROOT_URI_PATH)
```
Comment (suyograo) [inline lib/logstash/outputs/elasticsearch/http_client/pool.rb]: @andrewvc should `perform_request_to_url` use the safe URL?

File: spec/unit/outputs/elasticsearch/http_client/pool_spec.rb [hunk 2]
```diff
@@ -44,6 +44,25 @@
       expect(subject).to receive(:healthcheck!).once
       sleep(subject.resurrect_delay + 1)
     end
+
+    context "when using an absolute_healthcheck_path option" do
```
Comment (jsvd) [inline spec/unit/outputs/elasticsearch/http_client/pool_spec.rb]: I suggest using describe:

```ruby
describe "absolute_healthcheck_path" do
  context "when enabled" do
    # ...
  end
end
```

File: spec/unit/outputs/elasticsearch/http_client/pool_spec.rb [hunk 3]
```diff
@@ -44,6 +44,25 @@
...
+
+      before do
+        pool.start
+      end
+
+      after do
+        pool.close
+      end
+
+      it "should use the healthcheck_path as URL to do a health check" do
```
Comment (jsvd) [inline spec/unit/outputs/elasticsearch/http_client/pool_spec.rb]: look at this feature, I'd probably test that:

1) when `absolute_healthcheck_path` is enabled, `perform_request_to_url` is called with the `url` argument equal to the url constructed from `healthcheck_path`

2) when disabled, it behaves as before: appending the path to the existing hosts url (again observable in `perform_request_to_url`'s arguments

3) what happens when you pass a full URL and disable `absolute_healthcheck_path` (should raise exception)

4) what happens when you pass a path and enable `absolute_healthcheck_path` (should raise exception)

File: lib/logstash/outputs/elasticsearch/http_client/pool.rb [hunk 4]
```diff
@@ -225,9 +229,16 @@ def healthcheck!
       # Try to keep locking granularity low such that we don't affect IO...
       @state_mutex.synchronize { @url_info.select {|url,meta| meta[:state] != :alive } }.each do |url,meta|
         begin
-          logger.info("Running health check to see if an Elasticsearch connection is working",
-                      url: url, healthcheck_path: healthcheck_path)
-          response = perform_request_to_url(url, :head, healthcheck_path)
+          if @absolute_healthcheck_path
```
Comment (jsvd) [inline lib/logstash/outputs/elasticsearch/http_client/pool.rb]: I suggest deduping this into something like:

```ruby
          if @absolute_healthcheck_path
            url = absolute_healthcheck_safe_url.sanitized
            path = ROOT_URI_PATH
          else
            url = url # superfluous
            path = healthcheck_path
          end
          logger.info("Running health check to see if an Elasticsearch connection is working",
                        :url => url, :healthcheck_path => path)
          response = perform_request_to_url(url, :head, path)
```

File: lib/logstash/outputs/elasticsearch/http_client/pool.rb (hunk 1 above)
Comment (andrewvc) [inline lib/logstash/outputs/elasticsearch/http_client/pool.rb]: At this point everything uses the SafeURI. We only unwrap it briefly in the ManticoreAdapter *just* before passing it in to Manticore (which does not understand SafeURI)

File: lib/logstash/outputs/elasticsearch.rb [hunk 5]
```diff
@@ -176,7 +176,11 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
   # to service requests. If you have custom firewall rules you may need to change
   # this
   config :healthcheck_path, :validate => :string, :default => "/"
-
+  # When a `healthcheck_path` config is provided, this additional flag can be used to 
+  # specify whether it is an absolute URI or a relative path. 
+  # If this flag is true, `healthcheck_path` is expected to be a fully formed URL 
+  # with any basic auth credentials provided in the URL itself.
+  config :absolute_healthcheck_path, :validate => :boolean, :default => false
```
Comment (jsvd) [inline lib/logstash/outputs/elasticsearch.rb]: in order for this setting to reach the pool class, it is necessary to pass it down through the http_client_builder and http_client too.

Comment (suyograo) [inline lib/logstash/outputs/elasticsearch.rb]: Done

**Comment by suyograo** [comment]
@jsvd I wasn't able to get the tests updated.

**Comment by jsvd** [comment]
this seems to be failing because the top level `before :each` of this spec suite introduces a stub on the adapter that is getting in the way of the failed test.
it might be necessary to move [this](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb#L15) to localized spec sections (instead of global).

**Comment by jsvd** [comment]
closed in favor https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/538

**Pull request closed without merging**